fix!: fix valueHasUri bad values and missing types (DEV-1036)

mpro7 opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: DEV-1036


## What is the new behavior?


## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

subotic commented on webapi/src/main/scala/org/knora/webapi/store/triplestore/upgrade/plugins/UpgradePluginPR2079.scala:
```diff
@@ -0,0 +1,67 @@
+          case literal: DatatypeLiteral =>
+            if (literal.datatype != OntologyConstants.Xsd.Uri) {
+              statementsToRemove += statement
+
```
```suggestion
val newObjectValue = nodeFactory.makeDatatypeLiteral(literal.value, OntologyConstants.Xsd.Uri)
```

mpro7 replied:
Thanks for suggestion, but it needs to take parameter or, because in another case it takes differently named value to create `DatatypeLiteral`.

subotic commented on webapi/src/main/scala/org/knora/webapi/store/triplestore/upgrade/plugins/UpgradePluginPR2079.scala:
```diff
@@ -0,0 +1,67 @@
+              statementsToAdd += nodeFactory.makeStatement(
+                subj = statement.subj,
+                pred = statement.pred,
+                obj = nodeFactory.makeDatatypeLiteral(literal.value, OntologyConstants.Xsd.Uri),
```
```suggestion
                obj = newObjectValue,
```

subotic commented on webapi/src/main/scala/org/knora/webapi/store/triplestore/upgrade/plugins/UpgradePluginPR2079.scala:
```diff
@@ -0,0 +1,67 @@
+              )
+
+              log.info(
+                s"Transformed valueHasIri: $literal to ${nodeFactory
+                  .makeDatatypeLiteral(literal.value, OntologyConstants.Xsd.Uri)}."
```
```suggestion
                s"Transformed valueHasIri: $literal to $newObjectValue."
```

irinaschubert commented on .gitignore:
```diff
@@ -39,11 +40,11 @@ sipi/test
 
```
Yes, sorry, that was my fault in the last PR where I added `.trig` instead of `*.trig`

irinaschubert commented on webapi/src/test/scala/org/knora/webapi/store/triplestore/upgrade/plugins/UpgradePluginPR2079Spec.scala:
```diff
@@ -0,0 +1,84 @@
+      }
+    }
+
+    "fix value valueHasUri etered ad node w/o datatype" in {
```
what does this description mean? ;)

mpro7 replied:
Typos 😃

irinaschubert commented on webapi/src/test/scala/org/knora/webapi/store/triplestore/upgrade/plugins/UpgradePluginPR2079Spec.scala:
```diff
@@ -0,0 +1,84 @@
+      val plugin = new UpgradePluginPR2079(defaultFeatureFactoryConfig, log)
+      plugin.transform(model)
+
+      // Check that the value amd datatype was fixed.
```
```suggestion
      // Check that the value and datatype were fixed.
```

irinaschubert reviewed (approved):
LGTM, just some cosmetic suggestions. I would also recommend to double check if it works with standoff (but Ivan already mentioned that).

mpro7 commented on test_data/upgrade/pr2079.trig:
```diff
@@ -0,0 +1,29 @@
+    knora-base:standoffTagHasStartIndex     2 ;
+    knora-base:standoffTagHasStartParent     <http://rdfh.ch/0103/5LE8P57nROClWUxEPJhiug/values/fEbt5NzaSe6GnCqKoF4Nhg/standoff/1> ;
+    knora-base:standoffTagHasUUID           "dd2e785b-041e-4bcb-a77d-570efe8a068c" ;
+    knora-base:valueHasUri                  <http://www.maison-george-sand.fr/> .
```
I don't know how/where to check but in LS data there are only 2 instances of data written like in the code snippet above. Any other `valueHasUri` used in standoff looks like that:
```
knora-base:valueHasUri          "https://lumieres.unil.ch/fiches/bio/2666/"^^xsd:anyURI .
```

In our test data I found only one example which is typed string:
https://github.com/dasch-swiss/dsp-api/blob/main/test_data/all_data/anything-data.ttl#L1244

mpro7 commented:
@subotic I thought it is possible to run scripts on servers so this could be run like that -  manually. but if you say so I changed it to be the breaking change and updated version numbers.

BalduinLandolt commented on webapi/src/main/scala/org/knora/webapi/store/triplestore/upgrade/plugins/UpgradePluginPR2079.scala:
```diff
@@ -0,0 +1,68 @@
+    val newObjectValue: String => DatatypeLiteral = (value: String) =>
+      nodeFactory.makeDatatypeLiteral(value, OntologyConstants.Xsd.Uri)
+
+    for (statement: Statement <- model) {
```
this solution with a for-loop and two mutable collections seems very non-functional 😉

mpro7 replied:
Feel free to refactor it mate 💪

BalduinLandolt reviewed (approved):
LGTM
